src/Services/FHIR/FhirEncounterService.php

```diff
@@ -3,59 +3,141 @@
-        //$r = $this->createPractitionerResource($data['provider_id'], $temp);
-        $resource = new FHIREncounter();
+        return  [
+            '_id' => ['uuid'],
+            'patient' => ['pid'],
+            'date' => ['date']
+        ];
+    }
+
+    /**
+     * Parses an OpenEMR patient record, returning the equivalent FHIR Patient Resource
+     * https://build.fhir.org/ig/HL7/US-Core-R4/StructureDefinition-us-core-encounter-definitions.html
+     *
+     * @param array $dataRecord The source OpenEMR data record
+     * @param boolean $encode Indicates if the returned resource is encoded into a string. Defaults to false.
+     * @return FHIREncounter
+     */
+    public function parseOpenEMRRecord($dataRecord = array(), $encode = false)
+    {
+        $encounterResource = new FHIREncounter();
+
+        $meta = array('versionId' => '1', 'lastUpdated' => gmdate('c'));
+        $encounterResource->setMeta($meta);
+
         $id = new FhirId();
-        $id->setValue($eid);
-        $resource->setId($id);
-        $participant = new FHIREncounterParticipant();
-        $prtref = new FHIRReference();
-        $temp = 'Practitioner/' . $data['provider_id'];
-        $prtref->setReference($temp);
-        $participant->setIndividual($prtref);
-        $date = date('Y-m-d', strtotime($data['date']));
-        $participant->setPeriod(['start' => $date]);
-
-        $resource->addParticipant($participant);
-        $reason = new FHIRCodeableConcept();
-        $reason->setText($data['reason']);
-        $resource->addReasonCode($reason);
-        $resource->status = 'finished';
-        $resource->setSubject(['reference' => "Patient/$pid"]);
+        $id->setValue($dataRecord['uuid']);
+        $encounterResource->setId($id);
+
+        $status = new FHIRCode('finished');
+        $encounterResource->setStatus($status);
+
+        if (!empty($dataRecord['provider_id'])) {
+            $parctitioner = new FHIRReference(['reference' => 'Practitioner/' . $dataRecord['provider_id']]);
+            $participant = new FHIREncounterParticipant(array(
+                'individual' => $parctitioner,
+                'period' => ['start' => gmdate('c', strtotime($dataRecord['date']))]
+            ));
+            $encounterResource->addParticipant($participant);
+        }
+
+        if (!empty($dataRecord['reason'])) {
+            $reason = new FHIRCodeableConcept();
+            $reason->setText($dataRecord['reason']);
+            $encounterResource->addReasonCode($reason);
+        }
+
+        if (!empty($dataRecord['puuid'])) {
+            $patient = new FHIRReference(['reference' => 'Patient/' . $dataRecord['puuid']]);
+            $encounterResource->setSubject($patient);
+        }
+
+        $period = new FHIRPeriod();
+        $period->setStart(gmdate('c', strtotime($dataRecord['date'])));
+        $encounterResource->setPeriod($period);
+
+        // TODO:
+        // class
+        // type
+        // reasonCode (not mandatory)
```

Comment:
Still need to add Encounter.type & Encounter.class as we are following [US Core Profile](https://build.fhir.org/ig/HL7/US-Core-R4/StructureDefinition-us-core-encounter.html) and these are must haves so I think it will be manadatory to add these coding. I have discussed few things with brady sir but couldn't figure out which way would be better to implement. If anyone can help me figure out which field we can map to the valueset so that we can easily support this :)

cc @rhausam @dixonwhitmire